File: 234-palindrome-linked-list/palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {

    // first find out the middle

    let slow=fast=head;
    while(fast&&fast.next){
        slow=slow.next;
        fast=fast.next.next;
    }

    // reverse the second part
    let prev = null;
    let curr = slow;

    while(curr){
        let temp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = temp;
    }

    // checking for palindrome

    let firstList = head; // start node head
    let secondList = prev; // end node prev

    while(secondList){
        if(firstList.val != secondList.val){
            return false;
        }
        firstList = firstList.next;
        secondList = secondList.next
    }
    return true;

    
};

File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {

        // find out the middle using slow & fast pointer approach
        ListNode slow = head;
        ListNode fast = head;

        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }

        // reverse the second list
        // assume prev=null and curr = middle(slow)

        ListNode prev = null;
        ListNode curr = slow;

        while(curr!=null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        
        // check for palindrome
        ListNode firstList = head;
        ListNode secondList = prev;

        while(secondList!=null){
            if(firstList.val!=secondList.val){
                return false;
            }
            firstList = firstList.next;
            secondList = secondList.next;
        }
        return true;
    }
}